src/SFA.DAS.AssessorService.Application.Api/Services/OrganisationMergingService.cs

```diff
@@ -0,0 +1,280 @@
+﻿using SFA.DAS.AssessorService.Application.Interfaces;
+using SFA.DAS.AssessorService.Data;
+using SFA.DAS.AssessorService.Domain.Entities;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace SFA.DAS.AssessorService.Application.Api.Services
+{
+    public static class MergeOrganisationStatus
+    {
+        public const string InProgress = "In Progress";
+        public const string Approved = "Approved";
+        public const string Deleted = "Deleted";
+        public const string Reverted = "Reverted";
+    }
+
+    public static class ReplicationType
+    {
+        public const string Before = "Before";
+        public const string After = "After";
+    }
+
+    public class OrganisationMergingService : IOrganisationMergingService
+    {
+        private readonly AssessorDbContext _dbContext;
+
+        public OrganisationMergingService(AssessorDbContext dbContext)
+        {
+            _dbContext = dbContext;
+        }
+
+        public async Task<MergeOrganisation> MergeOrganisations(Organisation primaryOrganisation, Organisation secondaryOrganisation, Guid mergedByUserId, DateTime secondaryStandardsEffectiveTo)
+        {
+            if (null == primaryOrganisation) throw new ArgumentNullException(nameof(primaryOrganisation), "primaryOrganisation must be specified.");
+            if (null == secondaryOrganisation) throw new ArgumentNullException(nameof(secondaryOrganisation), "secondaryOrganisation must be specified.");
+
+            // Create the MergeOrganisation
+
+            var mergeOrganisation = CreateMergeOrganisations(
+                primaryOrganisation.EndPointAssessorOrganisationId, 
+                secondaryOrganisation.EndPointAssessorOrganisationId, 
+                mergedByUserId);
+
+            // Create the "Before" snapshot
+
+            CreateStandardsSnapshot(mergeOrganisation, primaryOrganisation.EndPointAssessorOrganisationId, "Before");
+            CreateStandardsSnapshot(mergeOrganisation, secondaryOrganisation.EndPointAssessorOrganisationId, "Before");
+
+            // Perform the merge.
+
+            MergeOrganisationStandardsAndVersions(
+                primaryOrganisation,
+                secondaryOrganisation,
+                mergedByUserId,
+                secondaryStandardsEffectiveTo);
+
+            // Create the "After" snapshot.
+
+            CreateStandardsSnapshot(mergeOrganisation, primaryOrganisation.EndPointAssessorOrganisationId, "After");
```

Comment:
Between the Primary and Secondary snapshot, we're somewhat "abusing" EF, the first snapshot creates the copy, the merge happens against all the objects tracked against the EF context, then when the secondary snapshot comes along, we're relying on it returning to us the objects in the context that have yet to be committed. Allowing us to create what the secondary snapshot will look like even before it's saved.

I don't think there is a problem with this, I'm just noting it here for discussion